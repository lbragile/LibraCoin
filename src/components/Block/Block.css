.block-and-chain {
  display: flex;
  align-items: center;
  min-width: 20rem;
}

.block-background {
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 1rem 2rem;
}

/* .block-background:hover {
  cursor: pointer;
  border: 1px solid black;
} */

.chain-link {
  font-size: 3rem;
  transform: translateX(-15px) rotate(45deg);
}
