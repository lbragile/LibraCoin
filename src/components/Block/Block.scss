$valid_bg: #b4ff9a;
$invalid_bg: #ff9a9a;

$valid_solution: #008000 !important;
$invalid_solution: #f00 !important;

.block {
  width: 20rem;
}

.valid-block {
  background-color: $valid_bg;
}

.invalid-block {
  background-color: $invalid_bg;
}

.valid-solution {
  color: $valid_solution;
}

.invalid-solution {
  color: $invalid_solution;
}

.show-trans-eye:hover {
  cursor: pointer;
}
