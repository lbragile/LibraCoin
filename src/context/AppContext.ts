import React from "react";
import { IAction, IState } from "../typings/AppTypes";

type TAppContext = { state: IState; dispatch: React.Dispatch<IAction> } | undefined;

export const AppContext = React.createContext<TAppContext>(undefined);
