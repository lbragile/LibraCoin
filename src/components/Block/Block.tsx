import React, { useState, useEffect, useContext } from "react";
import { Form, InputGroup } from "react-bootstrap";

import Statistics from "./Statistics";
import { IAction, IBlock, IState } from "../../typings/AppTypes";

import "./Block.css";
import { digestMessage } from "../../utils/conversion";
import { AppContext } from "../../context/AppContext";
import { ACTIONS } from "../../enums/AppDispatchActions";
import BlockTrans from "./BlockTrans";

export default function Block({ details }: { details: IBlock }): JSX.Element {
  const { state, dispatch } = useContext(AppContext) as { state: IState; dispatch: React.Dispatch<IAction> };

  const [solution, setSolution] = useState<string>("");
  const [timestamp, setTimestamp] = useState<number>(Date.now());
  const [merkleRoot, setMerkleRoot] = useState<string>(details.merkleRoot);
  const [isValid, setIsValid] = useState<boolean>(details.valid ?? true);
  const [showTrans, setShowTrans] = useState<boolean>(false);

  // update timestamp when solution is mined
  useEffect(() => setTimestamp(Date.now()), [solution]);

  async function updateBlockStatus(e: React.ChangeEvent<HTMLInputElement>): Promise<void> {
    const newRoot = e.target.value;
    const newHash = await digestMessage(details.index + details.prevHash + newRoot);
    setIsValid(newRoot === merkleRoot);
    setTimestamp(Date.now());
    setSolution(newHash);
    setMerkleRoot(newRoot);

    // after updating the block, propagate the changes
    await propagateBlockStatus(newHash, false, newRoot);
  }

  async function propagateBlockStatus(
    prevHash: string,
    skipFirstUpdate: boolean,
    newRoot?: string,
    timestamp = Date.now()
  ): Promise<void> {
    const index = details.index;

    for (let i = index; i < state.chain.length; i++) {
      const merkleRoot = newRoot && i === index ? newRoot : state.chain[i].merkleRoot;
      const currHash = i === index ? prevHash : await digestMessage(i + prevHash + merkleRoot);

      const newBlock = {
        index: i,
        timestamp,
        prevHash: i === index ? details.prevHash : prevHash,
        currHash,
        transactions: state.chain[i].transactions,
        merkleRoot,
        valid: skipFirstUpdate ? i === index : false,
      };

      prevHash = currHash; // next block's prevHash is this block's currHash

      dispatch({ type: ACTIONS.UPDATE_BLOCK, payload: { block: newBlock } });
    }
  }

  return (
    <div className="col">
      <div className={"block my-3 mx-1 p-2 rounded " + (isValid ? "valid-block" : "invalid-block")}>
        <Form>
          <InputGroup className="my-2">
            <InputGroup.Prepend>
              <InputGroup.Text>Index</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control type="number" defaultValue={details.index} disabled={true} />
          </InputGroup>

          <InputGroup className="my-2">
            <InputGroup.Prepend>
              <InputGroup.Text>Timestamp</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control type="number" defaultValue={solution ? timestamp : details.timestamp} disabled={true} />
          </InputGroup>

          <InputGroup className="my-2">
            <InputGroup.Prepend>
              <InputGroup.Text>Previous #</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control type="text" defaultValue={details.prevHash} disabled={true} />
          </InputGroup>

          <InputGroup className="my-2">
            <InputGroup.Prepend>
              <InputGroup.Text>Current #</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control type="text" defaultValue={solution ? solution : details.currHash} disabled={true} />
          </InputGroup>

          <InputGroup className="my-2">
            <InputGroup.Prepend>
              <InputGroup.Text>Merkle #</InputGroup.Text>
            </InputGroup.Prepend>
            {details.index === 0 ? (
              <Form.Control type="text" defaultValue={""} disabled={true} />
            ) : (
              <React.Fragment>
                <Form.Control
                  type="text"
                  value={merkleRoot}
                  onChange={(e: React.ChangeEvent<HTMLInputElement>) => updateBlockStatus(e)}
                />
                <InputGroup.Append>
                  <InputGroup.Text className="show-trans-eye" onClick={() => setShowTrans(!showTrans)}>
                    {showTrans ? "ðŸ™ˆ" : "ðŸ™‰"}
                  </InputGroup.Text>
                </InputGroup.Append>
              </React.Fragment>
            )}
          </InputGroup>
        </Form>

        <Statistics
          chain={true}
          solution={solution}
          setSolution={setSolution}
          isValid={isValid}
          setIsValid={setIsValid}
          propagateBlockStatus={propagateBlockStatus}
        />
      </div>

      {showTrans && <BlockTrans block={details} setMerkleRoot={setMerkleRoot} />}
    </div>
  );
}
