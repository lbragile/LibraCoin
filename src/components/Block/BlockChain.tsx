import React, { useContext } from "react";

import Block from "./Block";
import BlockTrans from "./BlockTrans";

import { AppContext } from "../../context/AppContext";
import { IAction, IState } from "../../typings/AppTypes";

import "./Block.scss";

export default function BlockChain(): JSX.Element {
  const { state } = useContext(AppContext) as { state: IState; dispatch: React.Dispatch<IAction> };

  return (
    <div className="mx-3 row flex-nowrap overflow-auto">
      {state.chain.map((block, i) => (
        <div className="block" key={i}>
          <Block block={block} />
          {block.showTrans && <BlockTrans block={block} />}
        </div>
      ))}
    </div>
  );
}
