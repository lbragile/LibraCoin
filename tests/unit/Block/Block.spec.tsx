import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import { AppContext } from "../../../src/context/AppContext";
import Block from "../../../src/components/Block/Block";
import { IState } from "../../../src/typings/AppTypes";

const { state, dispatch } = global;

describe("in preview mode", () => {
  it("renders correctly", () => {
    Date.now = jest.fn().mockReturnValueOnce(45678);

    const { asFragment } = render(
      <AppContext.Provider value={{ state, dispatch }}>
        <Block chain={false} merkleRoot={"abc"} index={0} />
      </AppContext.Provider>
    );

    expect(screen.getByRole("form", { name: /Block Form/i })).toHaveFormValues({
      index: state.preview.index,
      timestamp: 45678,
      prevHash: state.preview.prevHash,
      currHash: state.preview.currHash,
      merkle: "abc"
    });

    const index = screen.getByRole("spinbutton", { name: /Block Index/i });
    expect(index).toBeDisabled();
    expect(index).not.toBeRequired();

    const timestamp = screen.getByRole("spinbutton", { name: /Block Timestamp/i });
    expect(timestamp).toBeDisabled();
    expect(timestamp).not.toBeRequired();

    const prevHash = screen.getByRole("textbox", { name: /Block PrevHash/i });
    expect(prevHash).toHaveAttribute("readOnly");
    expect(prevHash).not.toBeRequired();

    const currHash = screen.getByRole("textbox", { name: /Block CurrHash/i });
    expect(currHash).toHaveAttribute("readOnly");
    expect(currHash).not.toBeRequired();

    const merkle = screen.getByRole("textbox", { name: /Block Merkle/i });
    expect(merkle).toHaveAttribute("readOnly");
    expect(merkle).not.toBeRequired();

    expect(screen.queryByLabelText(/Show Trans/i)).not.toBeInTheDocument();
    expect(screen.queryByRole("button", { name: /Add Block/i })).not.toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it("Shows 'Add Block' button when valid", () => {
    const ogState = JSON.parse(JSON.stringify(state)) as IState;
    ogState.preview.valid = true;

    const { asFragment } = render(
      <AppContext.Provider value={{ state: ogState, dispatch }}>
        <Block chain={false} merkleRoot={"abc"} index={0} />
      </AppContext.Provider>
    );

    expect(screen.getByRole("button", { name: /Add Block/i })).toHaveTextContent("Add Block");

    expect(asFragment()).toMatchSnapshot();
  });
});
