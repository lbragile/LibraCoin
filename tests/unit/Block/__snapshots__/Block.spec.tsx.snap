// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in blockchain mode 'Add Block' button is not present when block is invalid 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #ff9a9a;
}

.c2:hover {
  cursor: pointer;
  background-color: #c7e4ff;
}

.c1.c1.c1 {
  font-weight: bold;
}

<form
    aria-label="Block Form Invalid"
    class="c0 p-2 rounded"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Index
        </span>
      </div>
      <input
        aria-label="Block Index"
        class="form-control"
        disabled=""
        name="index"
        type="number"
        value="2"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Timestamp
        </span>
      </div>
      <input
        aria-label="Block Timestamp"
        class="form-control"
        disabled=""
        name="timestamp"
        type="number"
        value="1619827200000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Previous #
        </span>
      </div>
      <input
        aria-label="Block PrevHash"
        class="text-truncate form-control"
        name="prevHash"
        readonly=""
        type="text"
        value="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Current #
        </span>
      </div>
      <input
        aria-label="Block CurrHash"
        class="text-truncate form-control"
        name="currHash"
        readonly=""
        type="text"
        value="ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ"
      />
    </div>
    <div
      class="mt-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Merkle #
        </span>
      </div>
      <input
        aria-label="Block Merkle"
        class="text-truncate form-control"
        name="merkleRoot"
        readonly=""
        type="text"
        value="567.89ZForthTransactionAZ567.89A"
      />
      <div
        class="input-group-append"
      >
        <span
          aria-label="Show Trans"
          class="c2 input-group-text"
        >
          <svg
            aria-label="Contract Icon"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M.172 15.828a.5.5 0 0 0 .707 0l4.096-4.096V14.5a.5.5 0 1 0 1 0v-3.975a.5.5 0 0 0-.5-.5H1.5a.5.5 0 0 0 0 1h2.768L.172 15.121a.5.5 0 0 0 0 .707zM15.828.172a.5.5 0 0 0-.707 0l-4.096 4.096V1.5a.5.5 0 1 0-1 0v3.975a.5.5 0 0 0 .5.5H14.5a.5.5 0 0 0 0-1h-2.768L15.828.879a.5.5 0 0 0 0-.707z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`in blockchain mode 'Add Block' button is not present when block is valid 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #b4ff9a;
}

.c2:hover {
  cursor: pointer;
  background-color: #c7e4ff;
}

.c1.c1.c1 {
  font-weight: bold;
}

<form
    aria-label="Block Form"
    class="c0 p-2 rounded"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Index
        </span>
      </div>
      <input
        aria-label="Block Index"
        class="form-control"
        disabled=""
        name="index"
        type="number"
        value="1"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Timestamp
        </span>
      </div>
      <input
        aria-label="Block Timestamp"
        class="form-control"
        disabled=""
        name="timestamp"
        type="number"
        value="1619827200000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Previous #
        </span>
      </div>
      <input
        aria-label="Block PrevHash"
        class="text-truncate form-control"
        name="prevHash"
        readonly=""
        type="text"
        value="0000000000000000000000000000000000000000000000000000000000000000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Current #
        </span>
      </div>
      <input
        aria-label="Block CurrHash"
        class="text-truncate form-control"
        name="currHash"
        readonly=""
        type="text"
        value="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
      />
    </div>
    <div
      class="mt-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Merkle #
        </span>
      </div>
      <input
        aria-label="Block Merkle"
        class="text-truncate form-control"
        name="merkleRoot"
        readonly=""
        type="text"
        value="987.65EForthTransactionAE987.65A"
      />
      <div
        class="input-group-append"
      >
        <span
          aria-label="Show Trans"
          class="c2 input-group-text"
        >
          <svg
            aria-label="Expand Icon"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.828 10.172a.5.5 0 0 0-.707 0l-4.096 4.096V11.5a.5.5 0 0 0-1 0v3.975a.5.5 0 0 0 .5.5H4.5a.5.5 0 0 0 0-1H1.732l4.096-4.096a.5.5 0 0 0 0-.707zm4.344-4.344a.5.5 0 0 0 .707 0l4.096-4.096V4.5a.5.5 0 1 0 1 0V.525a.5.5 0 0 0-.5-.5H11.5a.5.5 0 0 0 0 1h2.768l-4.096 4.096a.5.5 0 0 0 0 .707z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`in blockchain mode handleViewTransactions changes emoji when show transactions is clicked 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #b4ff9a;
}

.c2:hover {
  cursor: pointer;
  background-color: #c7e4ff;
}

.c1.c1.c1 {
  font-weight: bold;
}

<form
    aria-label="Block Form"
    class="c0 p-2 rounded"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Index
        </span>
      </div>
      <input
        aria-label="Block Index"
        class="form-control"
        disabled=""
        name="index"
        type="number"
        value="1"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Timestamp
        </span>
      </div>
      <input
        aria-label="Block Timestamp"
        class="form-control"
        disabled=""
        name="timestamp"
        type="number"
        value="1619827200000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Previous #
        </span>
      </div>
      <input
        aria-label="Block PrevHash"
        class="text-truncate form-control"
        name="prevHash"
        readonly=""
        type="text"
        value="0000000000000000000000000000000000000000000000000000000000000000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Current #
        </span>
      </div>
      <input
        aria-label="Block CurrHash"
        class="text-truncate form-control"
        name="currHash"
        readonly=""
        type="text"
        value="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
      />
    </div>
    <div
      class="mt-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Merkle #
        </span>
      </div>
      <input
        aria-label="Block Merkle"
        class="text-truncate form-control"
        name="merkleRoot"
        readonly=""
        type="text"
        value="987.65EForthTransactionAE987.65A"
      />
      <div
        class="input-group-append"
      >
        <span
          aria-label="Show Trans"
          class="c2 input-group-text"
        >
          <svg
            aria-label="Contract Icon"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M.172 15.828a.5.5 0 0 0 .707 0l4.096-4.096V14.5a.5.5 0 1 0 1 0v-3.975a.5.5 0 0 0-.5-.5H1.5a.5.5 0 0 0 0 1h2.768L.172 15.121a.5.5 0 0 0 0 .707zM15.828.172a.5.5 0 0 0-.707 0l-4.096 4.096V1.5a.5.5 0 1 0-1 0v3.975a.5.5 0 0 0 .5.5H14.5a.5.5 0 0 0 0-1h-2.768L15.828.879a.5.5 0 0 0 0-.707z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`in blockchain mode renders correctly index === 0 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #b4ff9a;
}

.c1.c1.c1 {
  font-weight: bold;
}

<form
    aria-label="Block Form"
    class="c0 p-2 rounded"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Index
        </span>
      </div>
      <input
        aria-label="Block Index"
        class="form-control"
        disabled=""
        name="index"
        type="number"
        value="0"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Timestamp
        </span>
      </div>
      <input
        aria-label="Block Timestamp"
        class="form-control"
        disabled=""
        name="timestamp"
        type="number"
        value="1619827200000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Previous #
        </span>
      </div>
      <input
        aria-label="Block PrevHash"
        class="text-truncate form-control"
        name="prevHash"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Current #
        </span>
      </div>
      <input
        aria-label="Block CurrHash"
        class="text-truncate form-control"
        name="currHash"
        readonly=""
        type="text"
        value="0000000000000000000000000000000000000000000000000000000000000000"
      />
    </div>
    <div
      class="mt-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Merkle #
        </span>
      </div>
      <input
        aria-label="Block Merkle Genesis"
        class="form-control"
        disabled=""
        name="merkleRoot"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`in blockchain mode renders correctly index > 0 index → 1, class → {"background": "#b4ff9a"} 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #b4ff9a;
}

.c2:hover {
  cursor: pointer;
  background-color: #c7e4ff;
}

.c1.c1.c1 {
  font-weight: bold;
}

<form
    aria-label="Block Form"
    class="c0 p-2 rounded"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Index
        </span>
      </div>
      <input
        aria-label="Block Index"
        class="form-control"
        disabled=""
        name="index"
        type="number"
        value="1"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Timestamp
        </span>
      </div>
      <input
        aria-label="Block Timestamp"
        class="form-control"
        disabled=""
        name="timestamp"
        type="number"
        value="1619827200000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Previous #
        </span>
      </div>
      <input
        aria-label="Block PrevHash"
        class="text-truncate form-control"
        name="prevHash"
        readonly=""
        type="text"
        value="0000000000000000000000000000000000000000000000000000000000000000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Current #
        </span>
      </div>
      <input
        aria-label="Block CurrHash"
        class="text-truncate form-control"
        name="currHash"
        readonly=""
        type="text"
        value="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
      />
    </div>
    <div
      class="mt-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Merkle #
        </span>
      </div>
      <input
        aria-label="Block Merkle"
        class="text-truncate form-control"
        name="merkleRoot"
        readonly=""
        type="text"
        value="987.65EForthTransactionAE987.65A"
      />
      <div
        class="input-group-append"
      >
        <span
          aria-label="Show Trans"
          class="c2 input-group-text"
        >
          <svg
            aria-label="Expand Icon"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.828 10.172a.5.5 0 0 0-.707 0l-4.096 4.096V11.5a.5.5 0 0 0-1 0v3.975a.5.5 0 0 0 .5.5H4.5a.5.5 0 0 0 0-1H1.732l4.096-4.096a.5.5 0 0 0 0-.707zm4.344-4.344a.5.5 0 0 0 .707 0l4.096-4.096V4.5a.5.5 0 1 0 1 0V.525a.5.5 0 0 0-.5-.5H11.5a.5.5 0 0 0 0 1h2.768l-4.096 4.096a.5.5 0 0 0 0 .707z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`in blockchain mode renders correctly index > 0 index → 2, class → {"background": "#ff9a9a"} 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #ff9a9a;
}

.c2:hover {
  cursor: pointer;
  background-color: #c7e4ff;
}

.c1.c1.c1 {
  font-weight: bold;
}

<form
    aria-label="Block Form Invalid"
    class="c0 p-2 rounded"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Index
        </span>
      </div>
      <input
        aria-label="Block Index"
        class="form-control"
        disabled=""
        name="index"
        type="number"
        value="2"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Timestamp
        </span>
      </div>
      <input
        aria-label="Block Timestamp"
        class="form-control"
        disabled=""
        name="timestamp"
        type="number"
        value="1619827200000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Previous #
        </span>
      </div>
      <input
        aria-label="Block PrevHash"
        class="text-truncate form-control"
        name="prevHash"
        readonly=""
        type="text"
        value="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Current #
        </span>
      </div>
      <input
        aria-label="Block CurrHash"
        class="text-truncate form-control"
        name="currHash"
        readonly=""
        type="text"
        value="ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ"
      />
    </div>
    <div
      class="mt-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Merkle #
        </span>
      </div>
      <input
        aria-label="Block Merkle"
        class="text-truncate form-control"
        name="merkleRoot"
        readonly=""
        type="text"
        value="567.89ZForthTransactionAZ567.89A"
      />
      <div
        class="input-group-append"
      >
        <span
          aria-label="Show Trans"
          class="c2 input-group-text"
        >
          <svg
            aria-label="Contract Icon"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M.172 15.828a.5.5 0 0 0 .707 0l4.096-4.096V14.5a.5.5 0 1 0 1 0v-3.975a.5.5 0 0 0-.5-.5H1.5a.5.5 0 0 0 0 1h2.768L.172 15.121a.5.5 0 0 0 0 .707zM15.828.172a.5.5 0 0 0-.707 0l-4.096 4.096V1.5a.5.5 0 1 0-1 0v3.975a.5.5 0 0 0 .5.5H14.5a.5.5 0 0 0 0-1h-2.768L15.828.879a.5.5 0 0 0 0-.707z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`in preview mode renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #ff9a9a;
}

.c1.c1.c1 {
  font-weight: bold;
}

<form
    aria-label="Block Form Invalid"
    class="c0 col-12 col-lg-5 p-2 rounded"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Index
        </span>
      </div>
      <input
        aria-label="Block Index"
        class="form-control"
        disabled=""
        name="index"
        type="number"
        value="3"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Timestamp
        </span>
      </div>
      <input
        aria-label="Block Timestamp"
        class="form-control"
        disabled=""
        name="timestamp"
        type="number"
        value="1619913600000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Previous #
        </span>
      </div>
      <input
        aria-label="Block PrevHash"
        class="text-truncate form-control"
        name="prevHash"
        readonly=""
        type="text"
        value="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Current #
        </span>
      </div>
      <input
        aria-label="Block CurrHash"
        class="text-truncate form-control"
        name="currHash"
        readonly=""
        type="text"
        value="BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB"
      />
    </div>
    <div
      class="mt-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Merkle #
        </span>
      </div>
      <input
        aria-label="Block Merkle"
        class="text-truncate form-control"
        name="merkleRoot"
        readonly=""
        type="text"
        value="previewMerkle"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`in preview mode valid State Shows 'Add Block' button 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #b4ff9a;
}

.c2 {
  padding: 4px 0;
  margin-top: 8px;
  font-size: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1.c1.c1 {
  font-weight: bold;
}

<form
    aria-label="Block Form"
    class="c0 col-12 col-lg-5 p-2 rounded"
  >
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Index
        </span>
      </div>
      <input
        aria-label="Block Index"
        class="form-control"
        disabled=""
        name="index"
        type="number"
        value="3"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Timestamp
        </span>
      </div>
      <input
        aria-label="Block Timestamp"
        class="form-control"
        disabled=""
        name="timestamp"
        type="number"
        value="1619913600000"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Previous #
        </span>
      </div>
      <input
        aria-label="Block PrevHash"
        class="text-truncate form-control"
        name="prevHash"
        readonly=""
        type="text"
        value="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
      />
    </div>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Current #
        </span>
      </div>
      <input
        aria-label="Block CurrHash"
        class="text-truncate form-control"
        name="currHash"
        readonly=""
        type="text"
        value="BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB"
      />
    </div>
    <div
      class="mt-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="c1 input-group-text"
        >
          Merkle #
        </span>
      </div>
      <input
        aria-label="Block Merkle"
        class="text-truncate form-control"
        name="merkleRoot"
        readonly=""
        type="text"
        value="previewMerkle"
      />
    </div>
    <button
      aria-label="Add Block"
      class="c2 btn btn-success btn-block"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2 0a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2zm6.5 4.5v3h3a.5.5 0 0 1 0 1h-3v3a.5.5 0 0 1-1 0v-3h-3a.5.5 0 0 1 0-1h3v-3a.5.5 0 0 1 1 0z"
        />
      </svg>
    </button>
  </form>
</DocumentFragment>
`;
